File: frontend/src/lib/bout.ts
import genericRequest from "./requests";

export const HOME = 0;
export const AWAY = 1;

class Clock {
  public readonly startTimestamp: Date | null;
  public readonly elapsed: number;
  public readonly alarm: number;

  constructor(init?: Partial<Clock>) {
    Object.assign(this, init);
    if (this.startTimestamp !== null) {
      this.startTimestamp = new Date(this.startTimestamp);
    }
  }

  isRunning(): boolean {
    return this.startTimestamp !== null;
  }
}

class Timer {
  public readonly startTimestamp: Date | null;
  public readonly stopTimestamp: Date | null;
  public readonly period: number;
  public readonly jam: number;

  constructor(init?: Partial<Timer>) {
    Object.assign(this, init);
    if (this.startTimestamp !== null) {
      this.startTimestamp = new Date(this.startTimestamp);
    }
    if (this.stopTimestamp !== null) {
      this.stopTimestamp = new Date(this.stopTimestamp);
    }
  }

  hasStarted(): boolean {
    return this.startTimestamp !== null;
  }

  isRunning(): boolean {
    return this.hasStarted() && this.stopTimestamp === null;
  }
}

class Timeout extends Timer {
  constructor(init?: Partial<Timeout>) {
    super(init);
  }
}

export class Team {
  public readonly roster: { id: number; name: string };
  public readonly boutScore: number;
  public readonly jamScore: number;
  public readonly timeoutsRemaining: number;
  public readonly reviewsRemaining: number;
  public readonly scoreOffset: number;
}

export class Bout {
  public readonly id: number;
  public readonly ruleset: string;
  public readonly expectedStartTimestamp: Date | null;
  public readonly isFinal: boolean;
  public readonly jamCounts: number[];
  public readonly timeoutCounts: number[];
  public readonly clock: Clock;
  public readonly teams: Team[];
  public readonly activeJam: Timer | null;
  public readonly activeTimeout: Timeout | null;

  static generateKey(id: number) {
    return ["bouts", id];
  }

  constructor(init?: Partial<Bout>) {
    Object.assign(this, init);
    this.clock = new Clock(this.clock);
    if (this.expectedStartTimestamp !== null) {
      this.expectedStartTimestamp = new Date(this.expectedStartTimestamp);
    }
    if (this.activeJam !== null) {
      this.activeJam = new Timer(this.activeJam);
    }
    if (this.activeTimeout !== null) {
      this.activeTimeout = new Timeout(this.activeTimeout);
    }
    this.teams = this.teams.map<Team>((t) => Object.assign(new Team(), t));
  }

  getState(): "final" | "jam" | "lineup" | "stopped" | "timeout" {
    if (this.isFinal) {
      return "final";
    } else if (this.activeJam?.isRunning()) {
      return "jam";
    } else if (this.activeTimeout?.isRunning()) {
      return "timeout";
    } else if (this.activeJam !== null) {
      return "lineup";
    } else {
      return "stopped";
    }
  }

  async setupTrack(): Promise<void> {
    await genericRequest("rules/setup-track", "POST", { key: this.id });
  }

  async clearTrack(): Promise<void> {
    await genericRequest("rules/clear-track", "POST", { key: this.id });
  }

  async startJam(): Promise<void> {
    await genericRequest("rules/start-jam", "POST", { key: this.id });
  }

  async stopJam(): Promise<void> {
    await genericRequest("rules/stop-jam", "POST", { key: this.id });
  }

  async callTimeout(): Promise<void> {
    await genericRequest("rules/call-timeout", "POST", { key: this.id });
  }

  async endTimeout(): Promise<void> {
    await genericRequest("rules/end-timeout", "POST", { key: this.id });
  }
}

export interface BoutContext {
  jamDuration: number;
  lineupDuration: number;
  pointsPerTrip: number;
  numTimeouts: number;
  numReviews: number;
}

export async function getBout(boutId: number): Promise<Bout> {
  const response: Partial<Bout> = await genericRequest("bout", "GET", {
    boutId,
  });
  return new Bout(response);
}

export async function getBoutContext(boutId: number): Promise<BoutContext> {
  const response: BoutContext = await genericRequest("bout-context", "GET", {
    boutId,
  });
  return response;
}

File: backend/models/bout.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime  # noqa: TC003
from functools import cached_property
from typing import TYPE_CHECKING, Final, Literal, final

from core import (
    CHILD_RELATIONSHIP,
    PARENT_RELATIONSHIP,
    BaseSQLModel,
    CacheableSQLModel,
)
from sqlalchemy import Constraint, ForeignKey, UniqueConstraint, select
from sqlalchemy.orm import (
    Mapped,
    column_property,
    mapped_column,
    relationship,
)

from .jam import JamModel, TeamJamModel
from .time import ClockModel, TimeoutModel

if TYPE_CHECKING:
    from datetime import timedelta

    from .roster import RosterModel
    from .series import SeriesModel


REQUIRED_NUM_TEAMS: Final[int] = 2


@dataclass(frozen=True)
class BoutContext:
    jam_duration: timedelta
    lineup_duration: timedelta
    points_per_trip: int
    num_timeouts: int
    num_reviews: int


class GenericBoutModel(CacheableSQLModel):
    __tablename__: str = 'bouts'

    _series_id: Mapped[int] = mapped_column(ForeignKey('series.id'))
    _clock_id: Mapped[int] = mapped_column(ForeignKey('clocks.id', ondelete='RESTRICT'))
    expected_start_timestamp: Mapped[datetime | None] = mapped_column(default=None)
    is_final: Mapped[bool] = mapped_column(default=False)
    is_running: Mapped[bool] = mapped_column(default=False)
    order: Mapped[int] = mapped_column()
    ruleset: Mapped[str] = mapped_column()

    clock: Mapped[ClockModel] = relationship(
        cascade=CHILD_RELATIONSHIP,
        foreign_keys=[_clock_id],
        lazy='joined',
        single_parent=True,
    )
    jams: Mapped[list[JamModel]] = relationship(
        back_populates='bout',
        cascade=CHILD_RELATIONSHIP,
        lazy='selectin',
        order_by=[JamModel.period, JamModel.num],
    )
    series: Mapped[SeriesModel] = relationship(
        cascade=PARENT_RELATIONSHIP, foreign_keys=[_series_id], lazy='select'
    )
    teams: Mapped[list[GenericTeamModel]] = relationship(
        back_populates='bout',
        cascade=CHILD_RELATIONSHIP,
        lazy='selectin',
    )
    timeouts: Mapped[list[TimeoutModel]] = relationship(
        cascade=CHILD_RELATIONSHIP, lazy='selectin'
    )

    __table_args__: tuple[Constraint] = (UniqueConstraint(_series_id, order),)
    __mapper_args__: dict[str, str | bool] = {
        'polymorphic_abstract': True,
        'polymorphic_on': 'ruleset',
    }

    def __init__(
        self, series: SeriesModel, ruleset: str, *rosters: RosterModel
    ) -> None:
        if len(rosters) < REQUIRED_NUM_TEAMS:
            raise ValueError(f'A Bout must have at least {REQUIRED_NUM_TEAMS} Teams')
        order: int = 0 if len(series.bouts) == 0 else series.bouts[-1].order + 1
        super().__init__(
            series=series,
            order=order,
            clock=ClockModel(),
            ruleset=ruleset,
            teams=[GenericTeamModel(roster) for roster in rosters],
        )

    @final
    def get_state(self) -> Literal['final', 'jam', 'lineup', 'stopped', 'timeout']:
        if self.is_final:
            return 'final'
        if len(self.jams) > 0 and self.jams[-1].is_running():
            return 'jam'
        elif len(self.timeouts) > 0 and self.timeouts[-1].is_running():
            return 'timeout'
        elif self.is_running:
            return 'lineup'
        else:
            return 'stopped'

    def begin_period(self, timestamp: datetime) -> None: ...

    def end_period(self, timestamp: datetime) -> None: ...

    def start_jam(self, timestamp: datetime) -> None: ...

    def stop_jam(self, timestamp: datetime) -> None: ...

    def start_timeout(self, timestamp: datetime) -> None: ...

    def stop_timeout(self, timestamp: datetime) -> None: ...

    @cached_property
    def context(self) -> BoutContext: ...


class GenericTeamModel(BaseSQLModel):
    __tablename__: str = 'teams'

    _bout_id: Mapped[int] = mapped_column(ForeignKey('bouts.id'))
    _roster_id: Mapped[int] = mapped_column(ForeignKey('rosters.id'))
    reviews_remaining: Mapped[int] = mapped_column()
    score_offset: Mapped[int] = mapped_column(default=0)
    timeouts_remaining: Mapped[int] = mapped_column()

    ruleset = column_property(
        select(GenericBoutModel.ruleset)
        .where(GenericBoutModel.id == _bout_id)
        .scalar_subquery()
    )

    bout: Mapped[GenericBoutModel | None] = relationship(
        cascade=PARENT_RELATIONSHIP, lazy='selectin'
    )
    roster: Mapped[RosterModel | None] = relationship(
        cascade=PARENT_RELATIONSHIP, foreign_keys=[_roster_id], lazy='joined'
    )
    team_jams: Mapped[list[TeamJamModel]] = relationship(
        back_populates='team',
        cascade=CHILD_RELATIONSHIP,
        lazy='selectin',
        # order_by=[TeamJamModel.jam.period, TeamJamModel.jam.num], # FIXME
    )
    timeouts: Mapped[list[TimeoutModel]] = relationship(
        back_populates='team', cascade=CHILD_RELATIONSHIP, lazy='selectin'
    )

    __mapper_args__: dict[str, str | bool] = {
        'polymorphic_on': 'ruleset',
    }

    @classmethod
    def get_team_jam_score(cls, team_jam: TeamJamModel) -> int: ...

    def __init__(self, roster: RosterModel):
        super().__init__(roster=roster)

    @property
    def bout_score(self) -> int:
        bout_score: int = 0
        for team_jam in self.team_jams:
            bout_score += self.get_team_jam_score(team_jam)
        return bout_score

    @property
    def jam_score(self) -> int:
        if len(self.team_jams) == 0:
            return 0
        jam_score: int = self.get_team_jam_score(self.team_jams[-1])
        return jam_score
